### ISSUE TITLE
mixed_6/conv:0 layer in pytorch structure

### ISSUE THREAD

**Issue by wtliao** [openai/guided-diffusion#109]
Thanks for sharing the nice code. I want to use pytorch code for calculating the metrics. However, I dont know which layer in [pytorch InceptionV3](https://github.com/toshas/torch-fidelity/blob/154062cf937d54ed82bf01b59490dd3c21aa76c4/torch_fidelity/feature_extractor_inceptionv3.py#LL77C20-L77C20) is the one corresponding to the mixed_6/conv:0 layer in the tensorflow structure. Anyone has the idea? Thanks!

**Comment by wtliao** [comment]
I find the corresponding layer in the [pytorch InceptionV3](https://github.com/toshas/torch-fidelity/blob/154062cf937d54ed82bf01b59490dd3c21aa76c4/torch_fidelity/feature_extractor_inceptionv3.py#LL77C20-L77C20), it is the first 192 channels of [mixed_6d](https://github.com/toshas/torch-fidelity/blob/154062cf937d54ed82bf01b59490dd3c21aa76c4/torch_fidelity/feature_extractor_inceptionv3.py#L161).

But however, I found that, whatever the size of input image, the spatial_pred features are always (batch_size, 17, 17, 7). It seems that any input image is resized into 299x299? I am confused that, whether this resize operation can reflect the quality of generated image  not in 299x299 originally?

**Comment by lavinal712** [comment]
Hello, I am converting the evaluation metric code for ADM into PyTorch format. I wonder if you have finished writing the PyTorch version of the evaluation code.

[lavinal712/ADM-evaluation-suite-pytorch](https://github.com/lavinal712/ADM-evaluation-suite-pytorch)

```
$ python evaluator.py VIRTUAL_imagenet256_labeled.npz admnet_guided_upsampled_imagenet256.npz
...
computing reference batch activations...
computing/reading reference batch statistics...
computing sample batch activations...
computing/reading sample batch statistics...
Computing evaluations...
Inception Score: 1.100562334060669
FID: 3.943349109375731
sFID: 299.1543291933718
Precision: 0.82588
Recall: 0.5282
```

As you can see, there are issues with the sFID and Inception Score of the evaluation results. I myself am not very familiar with the operation of TensorFlow and Inception v3. If you are interested, you are welcome to review this code and suggest modifications.

**Comment by luowyang** [comment]
> Hello, I am converting the evaluation metric code for ADM into PyTorch format. I wonder if you have finished writing the PyTorch version of the evaluation code.
> 
> [lavinal712/ADM-evaluation-suite-pytorch](https://github.com/lavinal712/ADM-evaluation-suite-pytorch)
> 
> ```
> $ python evaluator.py VIRTUAL_imagenet256_labeled.npz admnet_guided_upsampled_imagenet256.npz
> ...
> computing reference batch activations...
> computing/reading reference batch statistics...
> computing sample batch activations...
> computing/reading sample batch statistics...
> Computing evaluations...
> Inception Score: 1.100562334060669
> FID: 3.943349109375731
> sFID: 299.1543291933718
> Precision: 0.82588
> Recall: 0.5282
> ```
> 
> As you can see, there are issues with the sFID and Inception Score of the evaluation results. I myself am not very familiar with the operation of TensorFlow and Inception v3. If you are interested, you are welcome to review this code and suggest modifications.

@lavinal712 I have found the reason: The TensorFlow version has a spatial feature shape of [None, 17, 17, 192] while the PyTorch version has [None, 192, 17, 17]. This will make great differences when the features are flattened. Doing a `permute` to the PyTorch version will align with the TF version and eliminate this issue.

Precisely, change [this line](https://github.com/lavinal712/ADM-evaluation-suite-pytorch/blob/1cc0cb8f63f106742a3c2dae9838390807b1368c/evaluator.py#L177) from
```python
spatial_pred = self.spatial_features.pop()[:, :7, :, :].detach().cpu().numpy()
```
to
```python
spatial_pred = self.spatial_features.pop().movedim(1, -1)[..., :7].detach().cpu().numpy()
```
will do the trick.

Edit: After some experiments, I found it also necessary to use the Inception model from [torch-fidelity](https://github.com/toshas/torch-fidelity) instead of [pytorch-fid](https://github.com/mseitzer/pytorch-fid). You can find a working version [here](https://github.com/luowyang/ADM-evaluation-suite-pytorch).
